src/librustdoc/html/render.rs

```diff
@@ -1602,13 +1620,26 @@ impl Context {
         } else {
             format!("{}-{}", item.source.loline, item.source.hiline)
         };
-        Some(format!(
-            "{root}src/{krate}/{path}#{lines}",
-            root = Escape(&root),
-            krate = krate,
-            path = path,
-            lines = lines
-        ))
+        if let Some(ref source_code_external_url) = self.shared.source_code_external_url {
+            Some(format!(
+                "{path}#{lines}",
+                path = compute_path(format!(
+                    "{root}{krate}/{path}",
```

Comment:
Hmm, I think this still won't work for the docs.rs use case because we don't have the crate name in the source. Example URL: https://docs.rs/crate/cranelift-faerie/0.59.0/source/src/lib.rs

The corresponding rustdoc URL is https://docs.rs/cranelift-faerie/0.59.0/src/cranelift_faerie/lib.rs.html#1-36, so I think we would pass `https://docs.rs/crate/cranelift-faerie/0.59.0/source/src/` as the external source root. This doesn't address @ollie27 's [point](https://github.com/rust-lang/rust/pull/69167#issuecomment-586732007) that the sources directory may not always be `src`, but at least it's a start.

Reply:
This was changed not to add `{krate}`.